src/TrimNoOps.cpp

```diff
@@ -0,0 +1,396 @@
+    }
+
+    void visit(const Min *op) {
+        if (!op->type.is_int() || op->type.bits() < 32) {
+            IRMutator::visit(op);
+        } else {
+            Expr a = mutate(op->a);
+            Expr b = mutate(op->b);
+            Expr test = a <= b;
+            if (provably_true_over_domain(a <= b)) {
+                expr = a;
+            } else if (provably_true_over_domain(b <= a)) {
+                expr = b;
+            } else {
+                expr = Min::make(a, b);
+            }
+        }
+    }
+
+    void visit(const Max *op) {
+        if (!op->type.is_int() || op->type.bits() < 32) {
+            IRMutator::visit(op);
+        } else {
+            Expr a = mutate(op->a);
+            Expr b = mutate(op->b);
+            if (provably_true_over_domain(a >= b)) {
+                expr = a;
+            } else if (provably_true_over_domain(b >= a)) {
+                expr = b;
+            } else {
+                expr = Max::make(a, b);
+            }
+        }
+    }
+
+    template<typename Cmp>
+    void visit_cmp(const Cmp *op) {
+        IRMutator::visit(op);
+        if (provably_true_over_domain(expr)) {
+            expr = make_one(op->type);
+        } else if (provably_true_over_domain(!expr)) {
+            expr = make_zero(op->type);
+        }
+    }
+
+    void visit(const LE *op) {
+        visit_cmp(op);
+    }
+
+    void visit(const LT *op) {
+        visit_cmp(op);
+    }
+
+    void visit(const GE *op) {
+        visit_cmp(op);
+    }
+
+    void visit(const GT *op) {
+        visit_cmp(op);
+    }
+
+    void visit(const EQ *op) {
+        visit_cmp(op);
+    }
+
+    void visit(const NE *op) {
+        visit_cmp(op);
+    }
+
+    template<typename StmtOrExpr, typename LetStmtOrLet>
+    StmtOrExpr visit_let(const LetStmtOrLet *op) {
+        Expr value = mutate(op->value);
+        containing_loops.push_back({op->name, {value, value}});
+        StmtOrExpr body = mutate(op->body);
+        containing_loops.pop_back();
+        return LetStmtOrLet::make(op->name, value, body);
+    }
+
+    void visit(const Let *op) {
+        expr = visit_let<Expr, Let>(op);
+    }
+
+    void visit(const LetStmt *op) {
+        stmt = visit_let<Stmt, LetStmt>(op);
+    }
+
+    void visit(const For *op) {
+        // Simplify the loop bounds.
+        Expr min = mutate(op->min);
+        Expr extent = mutate(op->extent);
+        containing_loops.push_back({op->name, {min, min + extent - 1}});
+        Stmt body = mutate(op->body);
+        containing_loops.pop_back();
+        stmt = For::make(op->name, min, extent, op->for_type, op->device_api, body);
+    }
+public:
+    SimplifyUsingBounds(const string &v, const Interval &i) {
+        containing_loops.push_back({v, i});
+    }
+
+    SimplifyUsingBounds() {}
+};
+
+class TrimNoOps : public IRMutator {
+    using IRMutator::visit;
+
+    void visit(const For *op) {
+        // TODO: bounds of GPU loops can't depend on outer gpu loop vars
```

Comment:
Need to take care of this TODO before we merge. I guess the simplest thing is to not attempt to trim gpu block or thread vars